### ISSUE TITLE
Can't send null over channel

### ISSUE THREAD

**Issue by withinboredom** [phasync/phasync#10]
A channel reader cannot receive a null value when used in a `foreach`, nor can you send one:

https://github.com/phasync/phasync/blob/5b5a81ea3da329ccebad2180341e1a86c52bd90b/src/Internal/ReadChannel.php#L38-L42

It might be a good idea to allow sending a `null`, but serialize to as a `SpecialValue::Null` enum type or something, then deserialize to `null` when reading from the channel.

Use case:

reading csvs which sometimes have a `null` in a column.

**Comment by frodeborli** [comment]
If I enable sending `null`, it has implications for the design of for example publisher/subscriber - making it slower. Also, I'm not sure how I would serialize `null` for example via a stream. A `null` value strictly speaking has a type (it's a pointer to the memory address 0, but it still has a type). In most cases, you would send an array of the entire CSV line which can contain null.

**Comment by withinboredom** [comment]
You can implement it like this via a dedicated type for it:

```php
enum SpecialValue { case NullValue; }
```

Then in the reader:

```php

public function getIterator(): Traversable { 
     while (null !== ($message = $this->read())) { 
         yield $message === SpecialValue::NullValue ? null : $message; 
     } 
 }
```

And in your writer:


```php
$value ??= SpecialValue::NullValue; // only set to null value if null already
```

**Comment by frodeborli** [comment]
The problem I have is that for every message, the PHP engine will have to check if the value is that special null representative. I will look into if I can implement this by actually sending the null value. If it has no performance impact, and the tests pass then that's fine.

**Comment by frodeborli** [comment]
Technically I am actually using the null value to close the channel. I carefully optimized performance so that I achieved 1 million messages per second. Supporting null significantly reduces that number, but sure an argument can be made that it is not neccesary. However, many things rely on channels, and often in many levels for a single operation and I spent time to get the performance to be absolute top. I will see if I can get the same performance with another approach.
